### PR TITLE
[v6] Update packet versions to the latest crypto refresh

### PR DESCRIPTION
**Pull request by twiss**
Implement the new packet versions from [draft-ietf-openpgp-crypto-refresh-08](https://datatracker.ietf.org/doc/html/draft-ietf-openpgp-crypto-refresh-08).

Fixes part of #1442; replaces #1505.

TODO:

- [x] update AEAD-encryption mechanism for private keys
- [x] cross-check with go-crypto
- [x] rebase on top of Ed25519 update to then add test vectors from RFC
- [x] add changes from https://github.com/twiss/openpgpjs/pull/3
- [x] add config to support v4 keys with legacy AEAD mode (from werner's draft) -- TODO in separate PR
- [x] properly squash commits

### CODE REVIEW

File: src/packet/sym_encrypted_integrity_protected_data.js [hunk 1]
```diff
@@ -113,33 +150,148 @@ class SymEncryptedIntegrityProtectedDataPacket {
...
+  /**
+   * En/decrypt the payload.
+   * NOTE: this function is also used by the AEADEncryptedDataPacket.
+   * @param {encrypt|decrypt} fn - Whether to encrypt or decrypt
+   * @param {Uint8Array} key - The session key used to en/decrypt the payload
+   * @param {Uint8Array | ReadableStream<Uint8Array>} data - The data to en/decrypt
+   * @returns {Promise<Uint8Array | ReadableStream<Uint8Array>>}
+   * @async
+   */
+  async aead(fn, key, data) {
```
Comment (larabr) [inline src/packet/sym_encrypted_integrity_protected_data.js:197]: Can this be a non-class function that simply takes a `packet: AEADEncryptedDataPacket | SymEncryptedIntegrityProtectedDataPacket` argument, and gets called by both classes? So we don't need to use `.apply` to reuse it. Also, doesn't `instanceof` work?

File: src/key/helper.js [hunk 2]
```diff
@@ -146,43 +146,58 @@ export async function getPreferredHashAlgo(key, keyPacket, date = new Date(), us
...
+      desiredCipherSuite :
+      defaultCipherSuite;
+  }
+  const defaultSymAlgo = enums.symmetric.aes128;
+  const desiredSymAlgo = config.preferredSymmetricAlgorithm;
+  return [
+    selfSigs.every(selfSig => selfSig.preferredSymmetricAlgorithms && selfSig.preferredSymmetricAlgorithms.includes(desiredSymAlgo)) ?
+      desiredSymAlgo :
+      defaultSymAlgo,
+    0
```
Comment (larabr) [inline src/key/helper.js:197]: Not a fan of this, I had to go and check whether `enums.aead.cfb === 0` since it got me confused; can we set `null` instead? Or just return `[<symAlgo>]`.
Based on the couple of calls I've seen, it's not clear to me whether we actually need an array -- can we can get away with `{ cipherAlgo, aeadAlgo }` ?

Comment (twiss) [reply src/key/helper.js]: :+1: I've changed it to `{ symmetricAlgo, aeadAlgo }`, with `undefined` meaning CFB.

File: src/packet/sym_encrypted_integrity_protected_data.js (hunk 1 above)
Comment (twiss) [reply src/packet/sym_encrypted_integrity_protected_data.js:197]: It's possible, but I don't want to change the code to accommodate the `AEADEncryptedDataPacket` use case too much; I would rather that the `SymEncryptedIntegrityProtectedDataPacket` is most easy to read, and `AEADEncryptedDataPacket` piggy-backs off of that. That way, if we can ever drop support for `AEADEncryptedDataPacket`, the code can stay essentially the same.
And, `instanceof` also works but it walks the entire prototype chain, which in this case I think isn't necessary since we don't have subclasses of `SymEncryptedIntegrityProtectedDataPacket` and `AEADEncryptedDataPacket`, so we can just check the constructor directly; that might also be slightly faster.

File: .eslintrc.js [hunk 3]
```diff
@@ -109,7 +109,7 @@ module.exports = {
     // Custom errors:
     'no-use-before-define': [2, { 'functions': false, 'classes': true, 'variables': false }],
     'no-constant-condition': [2, { 'checkLoops': false }],
-    'new-cap': [2, { 'properties': false, 'capIsNewExceptionPattern': 'EAX|OCB|GCM|CMAC|CBC|OMAC|CTR', 'newIsCapExceptionPattern': 'type|hash*' }],
+    'new-cap': [2, { 'properties': false, 'capIsNewExceptionPattern': 'EAX|OCB|GCM|CMAC|CBC|OMAC|CTR|HKDF', 'newIsCapExceptionPattern': 'type|hash*' }],
```
Comment (larabr) [inline .eslintrc.js:112]: Very minor but HKDF is really just a function that computes the hash, whereas all the others are somewhat class-like in that they return functions, so I wouldn't appease the linter and just rename the function to `computeHKDF` or the likes.

File: src/packet/public_key_encrypted_session_key.js [hunk 4]
```diff
@@ -67,13 +72,29 @@ class PublicKeyEncryptedSessionKeyPacket {
...
-    this.encrypted = crypto.parseEncSessionKeyParams(this.publicKeyAlgorithm, bytes.subarray(10));
+    if (this.version === 6) {
+      this.publicKeyVersion = bytes[offset++];
+      const fingerprintLength = this.publicKeyVersion ? (this.publicKeyVersion >= 5 ? 32 : 20) : 0;
+      this.publicKeyFingerprint = bytes.subarray(offset, offset + fingerprintLength);
+      offset += fingerprintLength;
+      if (this.publicKeyVersion === 0) {
+        this.publicKeyID = KeyID.wildcard();
+      } else if (this.publicKeyVersion >= 5) {
+        this.publicKeyID.read(this.publicKeyFingerprint);
```
Comment (lubux) [inline src/packet/public_key_encrypted_session_key.js:88]: Here it should be `this.publicKeyID.read(this.publicKeyFingerprint.subarray(-8));`
For v6 keys the Key ID is the high-order 64 bits of the fingerprint.

Comment (lubux) [reply src/packet/public_key_encrypted_session_key.js:88]: NVM is correct.

File: src/encoding/armor.js [hunk 5]
```diff
@@ -369,59 +273,51 @@ export function armor(messageType, body, partIndex, partTotal, customComment, co
...
       result.push('-----BEGIN PGP MESSAGE, PART ' + partIndex + '-----\n');
       result.push(addheader(customComment, config));
       result.push(base64.encode(body));
-      result.push('=', getCheckSum(bodyClone));
       result.push('-----END PGP MESSAGE, PART ' + partIndex + '-----\n');
       break;
     case enums.armor.signed:
       result.push('-----BEGIN PGP SIGNED MESSAGE-----\n');
-      result.push('Hash: ' + hash + '\n\n');
+      result.push(hash ? `Hash: ${hash} \n\n` : '\n');
```
Comment (twiss) [inline src/encoding/armor.js:292]:
```suggestion
      result.push(hash ? `Hash: ${hash}\n\n` : '\n');
```

File: src/message.js [hunk 6]
```diff
@@ -543,20 +549,28 @@ export class Message {
...
-      onePassSignatureList.forEach(onePassSig => packetlist.push(onePassSig));
+      onePassSignatureList.forEach(onePassSig => {
+        packetlist.push(onePassSig);
+        signatureSalts.push(onePassSig.salt);
+      });
     });
 
     packetlist.push(literalDataPacket);
-    packetlist.push(...(await createSignaturePackets(literalDataPacket, signingKeys, signature, signingKeyIDs, date, userIDs, notations, false, config)));
+    packetlist.push(...(await createSignaturePackets(literalDataPacket, signingKeys, signature, signingKeyIDs, date, userIDs, notations, signatureSalts, false, config)));
```
Comment (twiss) [inline src/message.js:569]: We could generate the salt when generating the signature packet instead, and then create the one-pass signature packet from the signature packet (though we could also do that as a separate change)

File: src/packet/sym_encrypted_integrity_protected_data.js [hunk 7]
```diff
@@ -54,28 +53,55 @@ class SymEncryptedIntegrityProtectedDataPacket {
   }
 
   constructor() {
-    this.version = VERSION;
+    this.version = 1;
```
Comment (larabr) [inline src/packet/sym_encrypted_integrity_protected_data.js:56]:
```suggestion
    this.version = null;
```
might have to be done in different PR

File: src/packet/signature.js [hunk 8]
```diff
@@ -509,7 +538,7 @@ class SignaturePacket {
         // Issuer Fingerprint
         this.issuerKeyVersion = bytes[mypos++];
         this.issuerFingerprint = bytes.subarray(mypos, bytes.length);
-        if (this.issuerKeyVersion === 5) {
+        if (this.issuerKeyVersion >= 5) {
```
Comment (larabr) [inline src/packet/signature.js:541]: Shouldn't we also throw if we read a issuerKeyID packet from a non-v4 key? There is such a key in the sop tests (unhashed packets). We risk that the keyID set here is overwritten, otherwise.

**Review by larabr — APPROVED** [review]
